I have chosen to code a Cournot compettition setting, where two firms compete each other on prices. For the standard model I have chosen firms with the same cost function and I will expand the model first by letting the firms face different cost functions, and later introducing a third firm, but then returning to same cost function.

Below the relevant packages for the first parts of the project is imported.

In [155]:
# Importing relevant packages

import numpy as np
import scipy as sp
import sympy as sm
import matplotlib.pyplot as plt

I start by defining the variables needed for the functions of the model

In [156]:
# Attaching symbols to relevant variables

q1 = sm.symbols('q_1')
q2 = sm.symbols('q_2') 
f = sm.symbols('f')
y = sm.symbols('y')

Below the basic functions of the model is defined.

In [161]:
# Definitions of the useful functions are made

# First off, the total demand function
def demand(q1,q2,y):
    return (y-2*(q1+q2))

# Secondly, I write the cost functions of Firm 1 and Firm 2:
def cost_1(q1,f):
    return (f+2*q1)

def cost_2(q2,f):
    return (f+2*q2)

# Finally, I establish the prfit function of first Firm 1 and then Firm 2:
def prof_1(q1,q2,y,f):
    return (demand(q1,q2,y) * q1 - cost_1(q1,f))
    
def prof_2(q1,q2,y,f):
    return (demand(q1,q2,y) * q2 - cost_2(q2,f))

To find the best response functions of the firms, the first order conditions are derived below

In [162]:
#Take the derivative of pi1 wrt q1 and the derivative of pi2 wrt q2 and set equal to 0 to get the first order condition

FOC1=sm.diff(prof_1(q1,q2,y,f),q1)
FOC2=sm.diff(prof_2(q1,q2,y,f),q2)

#Solve FOC1 for q1 to get the optimal choice of quantity, q1e, end FOC2 for q2 to get q2e
display(FOC1, FOC2)

-4*q_1 - 2*q_2 + y - 2

-2*q_1 - 4*q_2 + y - 2

In [163]:
# Now the best response functions are found by substitution and isolation of q_1 and q_2:

BestR_1= sm.solve(FOC1,q1)
BestR_2= sm.solve(FOC2,q2)
print("\n Firm 1 best response is: q_1 = " + str(BestR_1))
print("\n Firm 2 best response is: q_2 =" + str(BestR_2) )



 Firm 1 best response is: q_1 = [-q_2/2 + y/4 - 1/2]

 Firm 2 best response is: q_2 =[-q_1/2 + y/4 - 1/2]


In order to find the quantity each firm will produce, the response functions are substitued into the first order conditions

In [164]:
# The quantity of firm is found by substituting best response of b into FOC1 and vice versa 
BestR_1_s = FOC1.subs(q2,BestR_2)
opt_q1 = sm.solve(BestR_1_s,q1)

BestR_2_s = FOC2.subs(q1,BestR_1)
opt_q2 = sm.solve(BestR_2_s,q2)

# The optimal qunatities are shown
display (opt_q1, opt_q2) #Showing the optimal quantities

[-q_2/2 + y/4 - 1/2]

[-q_1/2 + y/4 - 1/2]

I will now find the price that the firms takes

In [165]:
#We are now able of finding the prices the firms will take for the good and the profit each firm get:

#As the above quantities is given as lists we start by defining/converting the list into a float:
p_q1=opt_q1[0]
p_q2=opt_q2[0]
#We can now find the price of the good:
print("\n The price is given by: " + str(demand(p_q1,p_q2,y)) ) 


 The price is given by: q_1 + q_2 + 2


As the last step of the analytically solution of the model, the profit for both firms are found

In [166]:
# Profit 1
prof_1(p_q1,p_q2,y,f)

# Profit 2
prof_2(p_q1,p_q2, y, f)

display (prof_1(p_q1,p_q2,y,f), prof_2(p_q1,p_q2,y,f))

-f + q_2 - y/2 + (q_1 + q_2 + 2)*(-q_2/2 + y/4 - 1/2) + 1

-f + q_1 - y/2 + (-q_1/2 + y/4 - 1/2)*(q_1 + q_2 + 2) + 1

Now the Numerical Analysis is done, and y is set to 50 and f is set to 10

I start by solving the first order conditions, best responses and finding the optimal quantities of the firms

In [167]:
# Setting parametervalues

y = 50
f = 10

# Finding first order conditions

FOC1 = sm.diff(prof_1(q1,q2,y,f),q1)
FOC2 = sm.diff(prof_2(q1,q2,y,f),q2)
display(FOC1,FOC2)

# Deriving best response functions 
BestR_1= sm.solve(sm.Eq(FOC1,0),q1)[0]
BestR_2= sm.solve(sm.Eq(FOC2,0),q2)[0]
print("\n The best repsonse of firm 1 is: q_1 = " + str(BestR_1) + " and the best repsonse of firm B is: q_2 =" + str(BestR_2) )

-4*q_1 - 2*q_2 + 48

-2*q_1 - 4*q_2 + 48


 The best repsonse of firm 1 is: q_1 = -q_2/2 + 12 and the best repsonse of firm B is: q_2 =-q_1/2 + 12


In [168]:
#Solving for the optimal quantity for each firm

BestR_2_s = FOC2.subs(q1,BestR_1) #substituting the best response into the FOC of firm A
opt_q2 = sm.solve(BestR_2_s,q2)

BestR_1_s = FOC1.subs(q2,BestR_2) #substituting the best response function of firm B into the FOC of firm A
opt_q1 = sm.solve(BestR_1_s,q1)

display (opt_q1, opt_q2)

[8]

[8]

As seen above the optimal quantities for both firm 1 and firm 2 is producing 8 units each

In [124]:
# FInding the price for the gooods produced

p_q1=opt_q1[0]
p_q2=opt_q2[0]

print("\n The price is: " + str(demand(p_q1,p_q2,y)) )


 The price is: 18


In [125]:
# Finally the profit of each firm is found by inserting the optimal values into the profit functions

print("\n Firm 1 profit is: " + str(prof_1(p_q1,p_q2,y,f)) + " and firm 2 profit is: " + str(prof_2(p_q1,p_q2,y,f)) )


 Firm 1 profit is: 118 and firm 2 profit is: 118


To conclude on the numerical analysis:
    - Firm 1:
        - Optimal quantity: 8
        - Profit 118
    - Firm 2:
        - Optimal quantity: 8
        - Profit 118
        
With a shared price for the good of 18

Below are the best repsonse functions graphed

In [126]:
# The two best response functions are plotted against each other

plt.plot([0,12],[24,0], label = 'BestR_1(opt_q2)')
plt.plot([0,24],[12,0], label = 'BestR_2(opt_q1)')
plt.legend()
plt.xlabel('Quantity firm 1')
plt.ylabel('Quantity firm 2')
plt.title('Best response functions')

print("\n The optimal point of quantities is " + str(p_q1) + " for firm 1 " + "and " + str(p_q2) + " for firm 2 " )


 The optimal point of quantities is 8 for firm 1 and 8 for firm 2 


Now I will expand this by letting the firms face different cost functions

This will follow the same procedure as before, defining basic function, finding FOC, deriving best repsonse functions and solving for optimal quantities and finally finding profit. First off the analytical solution

In [127]:
# Attachin new symbols to relevant variables

k = sm.symbols('k')
g = sm.symbols('g')

# As the demand function does not change, I start by defining the cost functions of Firm 1 and Firm 2:
def cost_1_new(q1,g):
    return (g+2*q1)

def cost_2_new(q2,g):
    return (g+2+8*q2)

# Secondly, I establish the profit function of first Firm 1 and then Firm 2:
def prof_1_new(q1,q2,k,g):
    return (demand(q1,q2,k) * q1 - cost_1_new(q1,g))
    
def prof_2_new(q1,q2,y,f):
    return (demand(q1,q2,k) * q2 - cost_2_new(q2,g))

In [128]:
# Now the first order conditions are found

FOC1 = sm.diff(prof_1_new(q1,q2,k,g),q1)
FOC2 = sm.diff(prof_2_new(q1,q2,k,g),q2)
display(FOC1,FOC2)

k - 4*q_1 - 2*q_2 - 2

k - 2*q_1 - 4*q_2 - 8

In [129]:
# The best response functions are now derived by

BestR_1= sm.solve(sm.Eq(FOC1,0),q1)[0]
BestR_2= sm.solve(sm.Eq(FOC2,0),q2)[0]

print("\n The best response function of firm 1 is: q_1 = " + str(BestR_1))
print("\n The best response function of firm 2 is: q_2 = " + str(BestR_2) )


 The best response function of firm 1 is: q_1 = k/4 - q_2/2 - 1/2

 The best response function of firm 2 is: q_2 = k/4 - q_1/2 - 2


In [130]:
# The quantities of each firm is found by substitution

BestR_1_s = FOC1.subs(q2,BestR_2)
opt_q1 = sm.solve(BestR_1_s,q1)

BestR_2_s = FOC2.subs(q1,BestR_1)
opt_q2 = sm.solve(BestR_2_s,q2)

display (opt_q1, opt_q2)

[k/6 + 2/3]

[k/6 - 7/3]

In [131]:
# The price of the good is found by
p_q1=opt_q1[0]
p_q2=opt_q2[0]
#We can now find the price of the good:
print("\n The price is: " + str(demand(p_q1,p_q2,k)) ) 


 The price is: k/3 + 10/3


In [132]:
# Profit 1
prof_1(p_q1, p_q2,k,g)

# Profit 2
prof_2(p_q1, p_q2, k, g)

display (prof_1_new(p_q1, p_q2, k, g), prof_2_new(p_q1, p_q2, k, g))

-g - k/3 + (k/6 + 2/3)*(k/3 + 10/3) - 4/3

-g - 4*k/3 + (k/6 - 7/3)*(k/3 + 10/3) + 50/3

FOr the Numerical analysis, the same values as before is given so k = 50 and g = 10

In [133]:
# Setting parametervalues

k = 50
g = 10

# Finding first order conditions

FOC1 = sm.diff(prof_1_new(q1,q2,k,g),q1)
FOC2 = sm.diff(prof_2_new(q1,q2,k,g),q2)
display(FOC1,FOC2)

# Deriving best response functions 
BestR_1= sm.solve(sm.Eq(FOC1,0),q1)[0]
BestR_2= sm.solve(sm.Eq(FOC2,0),q2)[0]
print("\n The best repsonse of firm 1 is: q_1 = " + str(BestR_1) + " and the best repsonse of firm B is: q_2 =" + str(BestR_2) )

-4*q_1 - 2*q_2 + 48

-2*q_1 - 4*q_2 + 42


 The best repsonse of firm 1 is: q_1 = -q_2/2 + 12 and the best repsonse of firm B is: q_2 =-q_1/2 + 21/2


In [134]:
# The optimal quantities of each firm are found by

BestR_1_s = FOC1.subs(q2,BestR_2)
opt_q1 = sm.solve(BestR_1_s,q1)

BestR_2_s = FOC2.subs(q1,BestR_1)
opt_q2 = sm.solve(BestR_2_s,q2)

display (opt_q1, opt_q2)

[9]

[6]

In [135]:
# The price of the good is found by
p_q1=opt_q1[0]
p_q2=opt_q2[0]
#We can now find the price of the good:
print("\n The price is: " + str(demand(p_q1,p_q2,k)) ) 


 The price is: 20


In [136]:
# Profit 1
prof_1(p_q1, p_q2,k,g)

# Profit 2
prof_2(p_q1, p_q2, k, g)

print("\n Firm 1 profit is: " + str(prof_1(p_q1, p_q2, k, g)) + " and firm 2 profit is: " + str(prof_2(p_q1, p_q2, k, g)) )


 Firm 1 profit is: 152 and firm 2 profit is: 98


To conclude on the numerical analysis:
    - Firm 1:
        - Optimal quantity: 9
        - Profit 152
    - Firm 2:
        - Optimal quantity: 6
        - Profit 98
        
With a shared price for the good of 20

In [137]:
# The two best response functions are plotted against each other

plt.plot([0,12],[24,0], label = 'BestR_1(opt_q2)')
plt.plot([0,21],[21/2,0], label = 'BestR_2(opt_q1)')
plt.legend()
plt.xlabel('Quantity firm 1')
plt.ylabel('Quantity firm 2')
plt.title('Best response functions')

print("\n The optimal point of quantities is " + str(p_q1) + " for firm 1 " + "and " + str(p_q2) + " for firm 2 " )


 The optimal point of quantities is 9 for firm 1 and 6 for firm 2 


The last expansion of the model in including another firm all three with different cost functions as defined below

In [138]:
# Attaching symbols to relevant variables

q3 = sm.symbols('q_3') #quantity produced by firm 3
p = sm.symbols('p')
d = sm.symbols('d')
c1 = sm.symbols('c_1') #unit cost for firm 1 
c2 = sm.symbols('c_2') #unit cost for firm 2 
c3 = sm.symbols('c_3') #unit cost for firm 3

# Defining basic functions of the model

# First off, the total demand function
def demand(q1,q2,q3,y):
    return (p-2*(q1+q2+q3))

# Secondly, I write the cost functions of Firm 1, Firm 2 and Firm 3
def cost_1_3(q1,d):
    return (d+c1*q1)

def cost_2_3(q2,d):
    return (d+c2*q2)

def cost_3_3(q3,d):
    return (d+c3*q3)

# Finally, I establish the profit function of first Firm 1, then Firm 2 and then Firm 3
def prof_1(q1,q2,q3,p,d):
    return (demand(q1,q2,q3,p) * q1 - cost_1_3(q1,d))
    
def prof_2(q1,q2,q3,p,d):
    return (demand(q1,q2,q3,p) * q2 - cost_2_3(q2,d))

def prof_3(q1,q2,q3,p,d):
    return (demand(q1,q2,q3,p) * q3 - cost_3_3(q3,d))

To start off, I derive the first order conditions and isolate q1, q2 and q3 respectively

In [151]:
# In order to calcluate the best response function, I find the first order conditions

FOC1 = sm.Eq(sm.diff(prof_1(q1,q2,q3,p,d),q1),0)
FOC2 = sm.Eq(sm.diff(prof_2(q1,q2,q3,p,d),q2),0)
FOC3 = sm.Eq(sm.diff(prof_3(q1,q2,q3,p,d),q3),0)

sol = sm.solve([FOC1,FOC2,FOC3],[q1,q2,q3])
display(sol)

{q_1: 13/2, q_2: 5, q_3: 11/2}

For later use I define a solution function depending on the cost of each firm

Now the numerical analysis starts and I have chosen cost values of:
    c1 = 3
    c2 = 6
    c3= 5

In [152]:
# Setting parametervalues

c1 = 3
c2 = 6
c3 = 5
p = 50
d = 10



In [153]:
# In order to calcluate the best response function, I find the first order conditions

FOC1 = sm.Eq(sm.diff(prof_1(q1,q2,q3,p,d),q1),0)
FOC2 = sm.Eq(sm.diff(prof_2(q1,q2,q3,p,d),q2),0)
FOC3 = sm.Eq(sm.diff(prof_3(q1,q2,q3,p,d),q3),0)

sol = sm.solve([FOC1,FOC2,FOC3],[q1,q2,q3])
display(sol)

{q_1: 13/2, q_2: 5, q_3: 11/2}

In [154]:
BestR_1 = sm.solve(FOC1,q1)
BestR_2 = sm.solve(FOC2,q2)
BestR_3 = sm.solve(FOC3,q3)

print("\n Firm 1 best response is: q_1 = " + str(BestR_1))
print("\n Firm 2 best response is: q_2 = " + str(BestR_2))
print("\n Firm 2 best response is: q_2 = " + str(BestR_3))


 Firm 1 best response is: q_1 = [-q_2/2 - q_3/2 + 47/4]

 Firm 2 best response is: q_2 = [-q_1/2 - q_3/2 + 11]

 Firm 2 best response is: q_2 = [-q_1/2 - q_2/2 + 45/4]


Through the assignment I have tried to show how Cournot competition changes when introducing different aspects. The results clearly show the chocices the firms stands with in competition with another firm. Settling on the optimal quantity heavily relies on the circumstances and choices made by competing firms.